examples/petsc/src/swarmutils.c

```diff
@@ -593,3 +608,213 @@ PetscErrorCode DMSwarmProjectFromSwarmToCells(DM dm_swarm, const char *field, Ve
+  PetscCall(CreateRestrictionFromPlex(ceed, dm_mesh, 0, NULL, 0, &elem_restr_u_mesh));
+  PetscCall(CreateRestrictionFromPlex(ceed, dm_coord, 0, NULL, 0, &elem_restr_x_mesh));
+
+  CeedElemRestrictionCreateVector(elem_restr_u_mesh, &data->x_ceed, NULL);
+  CeedElemRestrictionCreateVector(elem_restr_u_mesh, &data->y_ceed, NULL);
+
+  // Swarm objects
+  {
+    const PetscInt *cell_points;
+    IS              is_points;
+    Vec             X_ref;
+    CeedInt         num_elem;
+
+    PetscCall(DMSwarmCreateReferenceCoordinates(dm_swarm, &is_points, &X_ref));
+    PetscCall(DMGetDimension(dm_mesh, &dim));
+    CeedElemRestrictionGetNumElements(elem_restr_u_mesh, &num_elem);
+    CeedElemRestrictionGetNumComponents(elem_restr_u_mesh, &num_comp);
+
+    PetscCall(ISGetIndices(is_points, &cell_points));
+    PetscInt num_points = cell_points[num_elem + 1] - num_elem - 2;
+    CeedInt  offsets[num_elem + 1 + num_points];
+
+    for (PetscInt i = 0; i < num_elem + 1; i++) offsets[i] = cell_points[i + 1] - 1;
+    for (PetscInt i = num_elem + 1; i < num_points + num_elem + 1; i++) offsets[i] = cell_points[i + 1];
+    PetscCall(ISRestoreIndices(is_points, &cell_points));
+
+    // -- Points restrictions
+    CeedElemRestrictionCreateAtPoints(ceed, num_elem, num_points, num_comp, num_points * num_comp, CEED_MEM_HOST, CEED_COPY_VALUES, offsets,
+                                      &elem_restr_u_points);
+    CeedElemRestrictionCreateAtPoints(ceed, num_elem, num_points, dim, num_points * dim, CEED_MEM_HOST, CEED_COPY_VALUES, offsets,
+                                      &elem_restr_x_points);
+    CeedElemRestrictionCreateAtPoints(ceed, num_elem, num_points, q_data_size, num_points * q_data_size, CEED_MEM_HOST, CEED_COPY_VALUES, offsets,
+                                      &elem_restr_q_data_points);
+
+    // -- Points vectors
+    CeedElemRestrictionCreateVector(elem_restr_q_data_points, &q_data_points, NULL);
+
+    // -- Ref coordinates
+    {
+      PetscMemType       X_mem_type;
+      const PetscScalar *x;
+
+      CeedVectorCreate(ceed, num_points * dim, &x_ref_points);
+
+      PetscCall(VecGetArrayReadAndMemType(X_ref, (const PetscScalar **)&x, &X_mem_type));
+      CeedVectorSetArray(x_ref_points, MemTypeP2C(X_mem_type), CEED_COPY_VALUES, (CeedScalar *)x);
+      PetscCall(VecRestoreArrayReadAndMemType(X_ref, (const PetscScalar **)&x));
+    }
+
+    // Create Q data
+    {
+      CeedQFunction qf_setup;
+      CeedOperator  op_setup;
+
+      {
+        Vec                X_loc;
+        CeedInt            len;
+        const PetscScalar *x;
+
+        PetscCall(DMGetCoordinatesLocal(dm_mesh, &X_loc));
+        PetscCall(VecGetLocalSize(X_loc, &len));
+        CeedVectorCreate(ceed, len, &x_coord);
+
+        PetscCall(VecGetArrayRead(X_loc, &x));
+        CeedVectorSetArray(x_coord, CEED_MEM_HOST, CEED_COPY_VALUES, (CeedScalar *)x);
+        PetscCall(VecRestoreArrayRead(X_loc, &x));
+      }
```

Comment:
Note to investigate for here/Ratel - really we shouldn't need to copy since this CeedVector is destroyed at the end of this function. Could P2C and C2P it back. Marginal memory savings reusing the same array instead of copying

Reply:
That's certainly true here, though there are some places in Ratel which hold onto the `CeedVector` as a passive input (e.g. platens). Although we have discussed removing that in favor of a more precise subset of that vector in the past.

Reply:
I think its worth not copying here and anywhere else in the PETSc examples and Ratel that we have this pattern inside the same function call of
1) create vector
2) copy coords over
3) use Operator
4) destroy Operator + vector

Reply:
Oh, I already changed this in Ratel